puzzle-utils/words/dropquote.py

```diff
@@ -0,0 +1,193 @@
+    current_letters = []
+    new_word = True
+    while row < len(grid) and column < len(grid[0]):
+        if grid[row][column] == '-' and new_word == False:
+            if current_word.find(' ') != -1:
+                # Overlapped columns could produce a superset of the actual possible words
+                snapshot.words[word_index] = DropQuoteWords(current_letters[:], dropquote_search_words(current_word, current_letters))
+            
+            word_index += 1
+            current_word = ''
+            current_letters = []
+            new_word = True
+        elif grid[row][column] == ' ':
+            current_word += ' '
+            current_letters.append(letters[column])
+            new_word = False
+        else:
+            current_word += grid[row][column]
+            current_letters.append('')
+            new_word = False
+        
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+    
+    if current_word.find(' ') != -1:
+        # Overlapped columns could produce a superset of the actual possible words
+        snapshot.words[word_index] = DropQuoteWords(current_letters[:], dropquote_search_words(current_word, current_letters))
+
+    return snapshot
+
+def dropquote_search_words(word, letters):
+    pattern = '^'
+    for index in range(len(word)):
+        if word[index] == ' ':
+            pattern += f'[{letters[index]}]'
+        else:
+            pattern += word[index]
+    pattern += '$'
+    return search(pattern)
+
+def dropquote_apply_word(grid, letters, word_index, word):
+    """
+    Applies a word at the word index to the grid and removes the consumed letters.
+    >>> grid = ['    -', '  -  ', '     ']
+    >>> letters = ['gor', 'afo', 'mn', 'ete', 'hs']
+    >>> dropquote_apply_word(grid, letters, 2, 'thrones')
+    >>> print(grid)
+    ['    -', '  -th', 'rones']
+    >>> print(letters)
+    ['go', 'af', 'm', 'e', '']
+    """
+    row, column = dropquote_locate_word(grid, letters, word_index)
+    dropquote_apply_word_validate(grid, letters, word, row, column)
+    dropquote_apply_word_at_location(grid, letters, word, row, column)
+
+def dropquote_locate_word(grid, letters, word_index):
+    row = 0
+    column = 0
+    current_word_index = 0
+    new_word = True
+    while row < len(grid) and column < len(grid[0]):
+        if grid[row][column] == '-' and new_word == False:
+            current_word_index += 1
+            new_word = True
+        elif grid[row][column] != '-' and new_word == True:
+            new_word = False
+            if current_word_index == word_index:
+                break
+
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+    
+    if row >= len(grid) or column > len(grid[0]):
+        raise ValueError('The word index is greater than the number of words.')
+    
+    return row, column
+
+def dropquote_apply_word_validate(grid, letters, word, row, column):
```

Comment:
This seems like it could either be called directly from dropquote_apply_word_at_location or just be a part of that method